dev/core#3095 Permit setting of format_locale, prefer if set

eileenmcnaughton opened the pull request:
Overview
----------------------------------------
Many sites in Australia, Canada and NZ use 'en_US' as the site language - as there is no practical difference and if the language files are not present then there is no actual option. However, in CiviCRM 5.47 this is used for displaying currency too - and it is non optimal

Before
----------------------------------------
No option to change language (in some cases)
![image](https://user-images.githubusercontent.com/336308/156669775-340aa09a-f5f5-4a85-a7c8-f80644ed42d9.png)

And for my 'Canadian' site....

![image](https://user-images.githubusercontent.com/336308/156669942-1814bae3-b615-4fee-9c1f-5d8b0b6ae04b.png)



After
----------------------------------------
Can set format locale & once set....


![image](https://user-images.githubusercontent.com/336308/156669703-0dd90955-8075-400a-8819-3c403f97e74e.png)

![image](https://user-images.githubusercontent.com/336308/156669654-83f6a7f8-f835-45ba-ae62-f0313744b301.png)


Technical Details
----------------------------------------
This is a minimal patch to address https://lab.civicrm.org/dev/core/-/issues/3095
in time for 5.47. With this set it is possible to change the format locale to
English, Canada or English, Australian (erm & sorta NZ!)
and the currency will only be displayed before dollar amounts NOT of that
currency.

This should be enough to mitigate that regression feeling but missing are
1) fixing the admin form to hide irrelevant settings if format_locale is set
2) the psuedoconstant is cludgey - existing stuff doesn't seem to work so
I added a function - also - if we ARE going to use this option group we
should .... add NZ to it

Comments
----------------------------------------

eileenmcnaughton committed:
dev/core#3095 Permit setting of format_locale, prefer if set

This is a minimal patch to address https://lab.civicrm.org/dev/core/-/issues/3095
in time for 5.47. With this set it is possible to change the format locale to
English, Canada or English, Australian (but not NZ!)
and the currency will only be displayed before dollar amounts NOT of that
currency.

This should be enough to mitigate that regression feeling but missing are
1) fixing the admin form to hide irrelevant settings if format_locale is set
2) the psuedoconstant is cludgey - existing stuff doesn't seem to work so
I added a function - also - if we ARE going to use this option group we
should .... add NZ to it

eileenmcnaughton committed:
Fix the world

eileenmcnaughton commented:
@seamuslee001 @colemanw if we can get this into 5.47 I think it's non-regressive - the fiddly parts can follow

seamuslee001 commented on CRM/Core/I18n.php:
```diff
@@ -239,6 +239,26 @@ public static function languages($justEnabled = FALSE) {
+    // Hacking in for now since the is probably the most important use-case for
+    // money formatting in an English speaking non-US locale based on any reasonable
+    // metric.
+    $return[$value['en_NZ']] = ts('English - New Zealand');
```
@eileenmcnaughton I think this should be just 'en_NZ' ?

eileenmcnaughton replied:
oh yeah

seamuslee001 commented on CRM/Core/I18n.php:
```diff
@@ -239,6 +239,26 @@ public static function languages($justEnabled = FALSE) {
+  public static function getLocaleOptions(): array {
+    $values = CRM_Core_OptionValue::getValues(['name' => 'languages'], $optionValues, 'label', TRUE);
+    foreach ($values as $value) {
+      $return[$value['name']] = $value['label'];
```
also you seem to be populating an array that hasn't been set here

eileenmcnaughton replied:
Yeah - that's poor form for IDE friendliness but doesn't enotice - I've fixed it now

demeritcowboy commented on settings/Localization.setting.php:
```diff
@@ -409,6 +409,28 @@
+    'quick_form_type' => 'Select',
+    'html_type' => 'Select',
+    'html_attributes' => [
+      'multiple' => 0,
```
There's something wrong here it seems to let me select multiple, although it only saves one. But more importantly I can't clear the field value once it's set.

totten replied:
Yeah, that's weird. Simply undeclaring`multiple=>0` brings it back to a proper single-select, but then it doesn't display the default/null value nicely. Added 9d14b6f716d5db15fce70a1648c6c206f3557047 to address those two things together.

demeritcowboy commented on CRM/Core/I18n.php:
```diff
@@ -239,6 +239,27 @@ public static function languages($justEnabled = FALSE) {
+    // Hacking in for now since the is probably the most important use-case for
+    // money formatting in an English speaking non-US locale based on any reasonable
+    // metric.
+    $return['en_NZ'] = ts('English - New Zealand');
```
The other ones all have the country in brackets, BUT:

On the larger r-tech, it does feel a little wrong hacking NZ in to the list. It seems like it can just be added as a real option value but without needing translations. It won't appear in the site language dropdown, because there's a built-in check that translation files exist, but it would still be available to the format_locale etc code when you retrieve the list of options values.

What I mean is just remove the hack at line 259, and then add it as a real option value (for testing purposes I just did `insert into civicrm_option_value values (null, 84, 'English (New Zealand)', 'en', 'en_NZ', null, 0,0,45, null, 0,0,1,null,null,null,null,null);`). That seems to work to get the desired functionality. Of course, for real it would go in the xml file.

totten committed:
format_locale - Present single-value option. Fix display of null value.

Overview: Fix display of new setting in admin form.

Before: Field incorrectly displays as multi-value selection.

After: Field displays as single-value selection. The null value is specifically allowed.

totten committed:
(NFC) `en_NZ` label should look like other `en_*` labels

demeritcowboy commented:
Adding merge on pass from chat conversation https://chat.civicrm.org/civicrm/pl/c8otu63kgbnsmn76uths1jtdjo
